flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/orderedmultisetstate/OrderedMultiSetState.java

```diff
@@ -0,0 +1,173 @@
+import org.apache.flink.api.java.tuple.Tuple3;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.runtime.orderedmultisetstate.linked.LinkedMultiSetState;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Optional;
+
+/**
+ * This class represents an interface for managing an ordered multi-set state in Apache Flink. It
+ * provides methods to add, append, and remove elements while maintaining insertion order.
+ *
+ * <p>The state supports two types of semantics for adding elements:
+ *
+ * <ul>
+ *   <li><b>Normal Set Semantics:</b> Replaces an existing matching element with the new one.
+ *   <li><b>Multi-Set Semantics:</b> Appends the new element, allowing duplicates.
+ * </ul>
+ *
+ * <p>Removal operations are supported with different result types, indicating the outcome of the
+ * removal process, such as whether all elements were removed, the last added element was removed,
+ * or no elements were removed.
+ *
+ * @param <T> The type of elements stored in the state.
+ */
+@Internal
+@Experimental
+public interface OrderedMultiSetState<T> {
+
+    /**
+     * Add the given element using a normal (non-multi) set semantics: if a matching element exists
+     * already, replace it (the timestamp is updated).
+     */
+    SizeChangeInfo add(T element, long timestamp) throws Exception;
+
+    /** Add the given element using a multi-set semantics, i.e. append. */
+    SizeChangeInfo append(T element, long timestamp) throws Exception;
+
+    /** Get iterator over all remaining elements and their timestamps, in order of insertion. */
+    Iterator<Tuple2<T, Long>> iterator() throws Exception;
+
+    /** Tells whether any state exists (in the given key context). */
+    boolean isEmpty() throws IOException;
+
+    /**
+     * Remove the given element. If there are multiple instances of the same element, remove the
+     * first one in insertion order.
+     */
+    Tuple3<RemovalResultType, Optional<T>, SizeChangeInfo> remove(T element) throws Exception;
+
+    /** Clear the state (in the current key context). */
+    void clear();
+
+    /** Load cache. */
+    void loadCache() throws IOException;
+
+    /** Clear caches. */
+    void clearCache();
+
+    /** Removal Result Type. */
+    enum RemovalResultType {
+        /**
+         * Nothing was removed (e.g. as a result of TTL or not matching key), the result will not
+         * contain any elements.
+         */
+        NOTHING_REMOVED,
+        /** All elements were removed. The result will contain the last removed element. */
+        ALL_REMOVED,
+        /**
+         * The most recently added element was removed. The result will contain the element added
+         * before it.
+         */
+        REMOVED_LAST_ADDED,
+        /**
+         * An element was removed, it was not the most recently added, there are more elements. The
+         * result will not contain any elements
+         */
+        REMOVED_OTHER
```

Comment:
I am curious why nothing is returned in this case, this seems inconsistent with  REMOVED_LAST_ADDED which will return the  element added before it.